packages/grid/_modules_/grid/hooks/features/columnVisibility/useGridVisibleColumns.ts

```diff
@@ -0,0 +1,75 @@
+import * as React from 'react';
+import { GridComponentProps } from '../../../GridComponentProps';
+import { GridApiRef } from '../../../models/api/gridApiRef';
+import { useGridApiMethod, useGridApiOptionHandler } from '../../root';
+import { GRID_COLUMN_VISIBILITY_CHANGE } from '../../../constants';
+import { useGridSelector } from '../core/useGridSelector';
+import { useGridState } from '../core/useGridState';
+import { GridVisibleColumnsApi } from '../../../models/api/gridVisibleColumnsApi';
+import { gridColumnsMetaSelector, visibleGridColumnsSelector } from './gridVisibleColumnsSelector';
+import { GridColumnsMeta } from '../../../models/colDef/gridColDef';
+import { allGridColumnsSelector } from '../columns/gridColumnsSelector';
+
+export const useGridVisibleColumns = (
+  apiRef: GridApiRef,
+  props: Pick<GridComponentProps, 'onColumnVisibilityChange'>,
+) => {
+  const [, , forceUpdate] = useGridState(apiRef);
+
+  const columnsMeta = useGridSelector(apiRef, gridColumnsMetaSelector);
+  const visibleColumns = useGridSelector(apiRef, visibleGridColumnsSelector);
+  const allColumns = useGridSelector(apiRef, allGridColumnsSelector);
+
+  const setColumnVisibility = React.useCallback(
+    (field: string, isVisible: boolean) => {
+      const col = apiRef.current.getColumn(field);
+      const updatedCol = { ...col, hide: !isVisible };
+
+      apiRef.current.updateColumns([updatedCol]);
+      forceUpdate();
+
+      apiRef.current.publishEvent(GRID_COLUMN_VISIBILITY_CHANGE, {
+        field,
+        colDef: updatedCol,
+        api: apiRef,
+        isVisible,
+      });
+    },
+    [apiRef, forceUpdate],
+  );
+
+  const getVisibleColumns = React.useCallback(
+    () => visibleGridColumnsSelector(apiRef.current.getState()),
+    [apiRef],
+  );
+
+  const getColumnsMeta: () => GridColumnsMeta = React.useCallback(() => columnsMeta, [columnsMeta]);
+
+  const getColumnIndex = React.useCallback(
```

Comment:
That one is interesting. Not sure if it makes more sense to be here or in the `useGridColumns`?

Reply:
When splitting in two folders I had to move it due to circular imports (the selector from each file importing the selector from the other file)

But if we keep a single folder (ie an single selector file), the question is worth asking.

https://github.com/mui-org/material-ui-x/pull/2241#discussion_r679795569